### PR TITLE
fix: dataTree mutation

### LINKED ISSUE #14699: [Bug]-[30000]:Data truncated without any functional change in code. Evaluations using data from previous run?
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

We have the following JS object.
```
export default {
	data:()=>{
	return	Countries.data.map(country=>{
		country.ccn3= country.ccn3+1
		return country
	})
	}
}
```

Data from above is bound to a table as  `{{jsObject.data()}}`. Whenever you touch this field in property pane, the data is truncated. country.ccn3 changes as follows 233->2331->23311->233111 and so on.

Countries.Data is accurate, but worker is using the previous computation and keeps updating it.

I'm not sure if a user will ever run into it, but there could be a bug somewhere causing unnecessary side affects. 

### Steps To Reproduce

1. Create a data source with this api https://restcountries.com/v3.1/all
2. Crate a jsObject with function shared above.
3. Bind `{{jsObject.data()}}` to table data.
4. Touch the table data in property pane [add a space]

You will see country.ccn3 value change in each eval cycle.

### Public Sample App

https://release.app.appsmith.com/app/data-cloning-issue/countries-table-62b15b0e3446e7397e69d83e

### Version

Both, Latest.

### PR DESCRIPTION
**Pull request by Rishabh-Rathod**
## Description

In this PR we fix the mutation of dataTree which got enabled by changes [here](https://github.com/appsmithorg/appsmith/pull/11001)

Fixes #14699

- Bug fix (non-breaking change which fixes an issue)

## How Has This Been Tested?

Cypress Test case
- [x] Try mutating Api response data and verify mutation doesn’t work 
- [ ] https://github.com/appsmithorg/TestSmith/issues/2050

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by Rishabh-Rathod — COMMENTED** [review]
Add a cypress test case to test mutation doesn’t happen again.

**Comment by rohan-arthur** [comment]
tested #15877 - looks good

File: app/client/src/workers/evaluate.ts [hunk 1]
```diff
@@ -348,14 +350,15 @@ export function isFunctionAsync(
   resolvedFunctions: Record<string, any>,
   logs: unknown[] = [],
 ) {
+  const clonedDataTree = klona(dataTree);
```
Comment (ApekshaBhosale) [inline app/client/src/workers/evaluate.ts:353]: you can club this with createGlobal data btw? it's the same functionality @Rishabh-Rathod

**Review by ApekshaBhosale — CHANGES_REQUESTED** [review]

File: app/client/src/workers/evaluate.ts (hunk 1 above)
Comment (Rishabh-Rathod) [reply app/client/src/workers/evaluate.ts:353]: We will be removing `isFunctionAsync` code with [this task ](https://github.com/appsmithorg/appsmith/pull/15092) which i will pick next. 

Does it seem fine to skip refactor for now as we are anyways gonna remove this code?
 
 Let me know what are your thoughts on it.  @ApekshaBhosale

Comment (ApekshaBhosale) [reply app/client/src/workers/evaluate.ts:353]: sure! let's not refactor for now if we are taking it up immediately @Rishabh-Rathod

Comment (Rishabh-Rathod) [reply app/client/src/workers/evaluate.ts:353]: @ApekshaBhosale Could you also review the PR as `request changes` blocks the PR until approved by same reviewer.

**Review by ApekshaBhosale — APPROVED** [review]

**Pull request merged**